[1.21.11] Migrate the MDK to ForgeGradle 7

Jonathing opened the pull request:
This PR migrates the MDK to ForgeGradle 7, and in the process, strips it of all complex tasks and build setup.

Complex MDKs will be created and preserved in the new [MDKExamples](https://github.com/MinecraftForge/MDKExamples). The top of the `build.gradle` file includes a link to this repository. I'm still unsure of what to do with `README.txt`.

Jonathing committed:
Use external Gradle wrapper for the MDK

Jonathing committed:
Fully trim down MDK, use ForgeGradle 7

LexManos commented on mdk/build.gradle:
```diff
@@ -1,235 +1,61 @@
-        server {
-            property 'forge.enabledGameTestNamespaces', mod_id
+        register('server') {
+            systemProperty 'forge.enabledGameTestNamespaces', mod_id
```
Move this to the configureEach.

LexManos commented on mdk/build.gradle:
```diff
@@ -1,235 +1,61 @@
-    // See https://docs.minecraftforge.net/en/latest/advanced/accesstransformers/ for more information.
-    // accessTransformer = file('src/main/resources/META-INF/accesstransformer.cfg')
+minecraft {
+    mappings channel: 'official', version: '@MC_VERSION@'
```
Could use a link to an example of changing to parchment.

Jonathing replied:
I will cover Parchment in ForgeGradle 7's documentation as it is natively supported and no longer needs Librarian.

LexManos commented on mdk/README.txt:
```diff
@@ -1,3 +1,4 @@
+TODO [Forge][MDK] Rework or trim down this entire document
```
"This example MDK is a simple showcase for the most basic of mods, to see more examples see the MDKExamples repo" or something like that. Point is we need to trim down a ton of the comments and docs because people simply dont read them.

LexManos replied:
Needs update

LexManos commented:
Its nice to see how small the buildscript can be with the unneeded things removed.
I'll have to get this on the mac/desktop to test but as for the code changes it looks fine to me. 
How much commenting do we want to do with the new build script? Having too much is obviously an issue. But we need to at least point people where to find more info.

Jonathing commented:
> How much commenting do we want to do with the new build script?

Parchment should get its own comment.

AccessTransformers and Mixin need to be mentioned in a way that points the user to MDKExamples.

Everything else is fair game. The comments I left in are the ones I feel justified in keeping.

PaintNinja commented on mdk/src/main/resources/META-INF/mods.toml:
```diff
@@ -1,73 +1,36 @@
+
+modLoader="javafml"
+loaderVersion="[@FORGE_SPEC_VERSION@,)"
 license="${mod_license}"
```
This references a deleted gradle property and is therefore an invalid licence

Jonathing replied:
Good catch, thank you.

PaintNinja commented on mdk/gradle.properties:
```diff
@@ -1,66 +1,9 @@
-# Sets default memory used for Gradle commands. Can be overridden by user or command line properties.
-# This is required to provide enough memory for the Minecraft decompilation process.
-org.gradle.jvmargs=-Xmx5G
```
Does removing this cause any problems for FG7?

Jonathing replied:
No, Mavenizer happens in an entirely forked process, which itself forks Java and Javac for the invoked tools. If you believe there might be OOM concerns, please open an issue for Mavenizer itself.

PaintNinja commented on mdk/gradle.properties:
```diff
@@ -1,66 +1,9 @@
-# Sets default memory used for Gradle commands. Can be overridden by user or command line properties.
-# This is required to provide enough memory for the Minecraft decompilation process.
-org.gradle.jvmargs=-Xmx5G
-org.gradle.daemon=false
```
While Gradle's default has been true for a while now, the Forge MDK has set it to false for a very long time. Removing it's fine, but I do wonder if explicitly setting to true would help mod devs migrating from FG6 know that it's now supported and recommended in FG7.

LexManos replied:
May be better to mention the gradle daemon in the migration guide. Specifically a quick history on why we disabled it and why the new system allows us to enable it again.

PaintNinja commented on mdk/src/main/resources/META-INF/mods.toml:
```diff
@@ -1,73 +1,36 @@
+#displayTest="MATCH_VERSION" #optional, default=MATCH_VERSION when not clientSideOnly, else IGNORE_ALL_VERSION
+
+description='''Example mod description.
+Newline characters can be used and will be replaced properly.'''
```
Replaced with what? With the processResources gone, it might be clearer to say something along the lines that the description supports a multiline string like this one.

Jonathing replied:
For the record, the complex examples in MDKExamples are going to be superseding what this will be. So I'm perfectly fine with just making this a one-line string or something.

Jonathing committed:
Fix issues with project import and running the game

Jonathing committed:
Add additional clarifications and rewordings

Jonathing committed:
Automate replacing required EventBus version

Jonathing committed:
Update ForgeGradle to 7.0.3

Jonathing committed:
Update MDK Gradle to 9.3.1

Jonathing committed:
Uncomment license field in mods.toml

Jonathing commented:
This is now ready for a re-review. I've reworded a bit of the English in the MDK to avoid as much potential confusion as possible. But I don't want to hand-hold people in this extremely minimal MDK, only meant to get the ball rolling. I've put at the top of the `build.gradle` where to find complex examples, to which **we will continue to add to.**

Updated to latest Gradle and ForgeGradle, removed Forge maven from plugin repositories, etc. I've tested the resultant MDK locally and everything works as intended.

LexManos commented on mdk/settings.gradle:
```diff
@@ -1,13 +1,7 @@
-}
-
+// NOTE: When ForgeGradle 7 is stabilized, it will be uploaded to the Gradle Plugin Portal.
+//       This block will be removed when that happens.
```
Needs removal

PaintNinja commented on mdk/settings.gradle:
```diff
@@ -1,13 +1,7 @@
-}
-
+// NOTE: When ForgeGradle 7 is stabilized, it will be uploaded to the Gradle Plugin Portal.
+//       This block will be removed when that happens.
```
This note can be removed now

Jonathing replied:
Done

LexManos commented:
Couple reminders, but besides that looks fine.

PaintNinja commented on mdk/build.gradle:
```diff
@@ -1,235 +1,58 @@
+// This is a minimal setup that works on all IDEs and environments without compromise.
+// For complex examples (including AccessTransformers, Jar-in-Jar, and Mixin), see our examples repository.
+// https://github.com/MinecraftForge/MDKExamples
```
Wording implies that the other examples have issues with IDE support.

Jonathing replied:
```suggestion
// This is a minimal setup that works on all IDEs and environments.
// For complex examples (including AccessTransformers, Jar-in-Jar, and Mixin), see our examples repository.
// https://github.com/MinecraftForge/MDKExamples
```

?

PaintNinja replied:
In before someone complains about NetBeans :P 

But yeah that works, or even just shorten to "This is a minimal build setup"

Jonathing replied:
jGRASP

Jonathing committed:
Edit and ignore mdk README.txt

Jonathing committed:
Removed testing comment

Jonathing committed:
Link to MDKExamples in the README.txt

Jonathing commented:
I've gone ahead and slightly edited the MDK's `README.txt` to remove legacy parts of the process and change what has changed in recent times. It will also be ignored by default in the MDK's git repo.

This does make your video on setting up the MDK outdated, but I am happy to make my own at a later time. Or you can, whoever does it first. I've also linked the MDKExamples repo in the README. I hope people still read it at least in part.

Jonathing commented:
Final pass then ship it.

PaintNinja commented on mdk/build.gradle:
```diff
@@ -1,235 +1,58 @@
-    }
-}
-
-tasks.withType(JavaCompile).configureEach {
-    options.encoding = 'UTF-8' // Use the UTF-8 charset for Java compilation
-}
```
Non-English mods and javadocs that use emojis or ascii art may break without this. I know newer Java defaults to it, but not sure if Gradle changes defaults to the older encoding regardless.

Jonathing replied:
[Default since Java 18](<https://openjdk.org/jeps/400>), and Gradle does not change encoding if it is not instructed to.

But sure. Since we'd need to add this in for the older MDKs (1.20.4-), we might as well keep it here for consistency. Good call.

PaintNinja replied:
The comment can be stripped at least, I dunno what I was thinking when I wrote that years ago... doesn't really tell you anything the code doesn't already say

Jonathing replied:
I made a commit which elaborates on why it's there, so we're good.

PaintNinja replied:
If definitely not needed on 1.20.6+ then I'm fine with it being removed and only being present on older MDKs, this is a minimal MDK after all. Or keep it for consistency, I'm fine either way. `¯\_(ツ)_/¯`

Jonathing committed:
javac with UTF-8 by default

Jonathing committed:
Reword top build.gradle comment to avoid confusion on MDK compatibility

PaintNinja commented on mdk/build.gradle:
```diff
@@ -1,235 +1,64 @@
-}
-
-// IntelliJ no longer downloads javadocs and sources by default, this tells Gradle to force IntelliJ to do it.
-idea.module { downloadJavadoc = downloadSources = true }
```
Maybe keep this and the Eclipse equiv? I think the eclipse equiv is `eclipse.classpath { download... = true }` off the top of my head, check traditional MDK in examples

Jonathing replied:
nahhh it'll be fine. IntelliJ is smart enough to attach sources (when asked to) nowadays and Eclipse users are built different from the rest of us

PaintNinja commented on mdk/build.gradle:
```diff
@@ -1,235 +1,64 @@
-        // applies to all the run configs below
         configureEach {
-            workingDirectory project.file('run')
+            workingDir = layout.projectDirectory.dir('run')
```
Minor: IntelliJ seems to ignore this in the linter due to mistreating the Gradle Internal annotation iirc. It also gets confused about where the layout property comes from. Using it.workingDir and this.layout seems to fix these two linting errors, but I haven't checked if it works the same yet (it should, but you never know with Gradle)

Jonathing replied:
IntelliJ issue. The DSL syntax is correct here. I'm sure in Kotlin this is linted properly as well.

PaintNinja commented:
Should the LICENCE.txt and CREDITS.txt still be kept in minimal MDKs?

LexManos commented:
License is needed for it to actually be a license tho could be renamed forge_ and gitiginored like the readme is. Credits can be streamlined/updated into the readme, just a basic 'thanks to everyone who has contributed and please consider helping out at <githublink>' or something like that.

LexManos commented:
Because this PR is meant to turn the MDK into a bare bones example. I have moved the Multi-AccessTransformer example to [MDKExamples](https://github.com/MinecraftForge/MDKExamples/commit/59deaa24011b6020b885d1b5326a61cd8d610dac)

LexManos commented:
Merged in https://github.com/MinecraftForge/MinecraftForge/commit/d29524bac380b13b4dd67f3fd48bef3524560664
